Update DocuSign REST API with OpenAPI v2.1

un3481 opened the pull request:
Some APIs have changed, so this PR will not be backwards compatible.

Package version should probably be changed from 0.4.4 to 0.5.

frahugo commented on mix.exs:
```diff
@@ -39,10 +39,11 @@ defmodule DocuSign.MixProject do
     [
-      {:poison, "~> 5.0"},
+      {:tesla, "~> 1.4"},
+      {:poison, "~> 3.0"},
```
Any reason why `poison` was downgraded from 5.0 to 3.0?

un3481 replied:
I used the version in the new generated app because I believed at the time that it would be higher than the current one. Now I see that only "tesla" was upgraded (to version 1.4) and poison had a downgrade. I will change poison version back to 5.0.

frahugo commented:
@un3481 In the IEX console, are you able to get a list of your templates in the sandbox? In the console, I'm able to establish a connection the API endpoint, but when I get the list of my templates in Docusign, I get a `404 - File or directory not found` error back. I need to spend more time and see why this is happening, Maybe the version of Poison? Or something else? 

Is it working with you? I use the same private key file as with v2.0 API, and same user ID and account ID. Everything works fine with v2.0.

```
> export DOCUSIGN_CLIENT_ID="xxxxxxx"
> iex -S mix
iex> user_id = "xxxxxxx"
iex> {:ok, connection} = DocuSign.Connection.get(user_id)
...
iex> account_id = "xxxxxx"
iex> {:ok, templates} = DocuSign.Api.Templates.templates_get_templates(connection, account_id, include: "recipients")
** (MatchError) no match of right hand side value: {:error, %Tesla.Env{__client__: %Tesla.Client{adapter: {Tesla.Adapter.Mint, :call, [[]]}, fun: nil, post: [], pre: [{Tesla.Middleware.BaseUrl, :call, ["https://demo.docusign.net"]}, {Tesla.Middleware.Headers, :call, [[{"authorization", "Bearer OBFUSCATED"}]]}, {Tesla.Middleware.EncodeJson, :call, [[engine: Poison]]}]}, __module__: DocuSign.Connection.Request, body: "<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\">\r\n<html xmlns=\"http://www.w3.org/1999/xhtml\">\r\n<head>\r\n<meta http-equiv=\"Content-Type\" content=\"text/html; charset=iso-8859-1\"/>\r\n<title>404 - File or directory not found.</title>\r\n<style type=\"text/css\">\r\n<!--\r\nbody{margin:0;font-size:.7em;font-family:Verdana, Arial, Helvetica, sans-serif;background:#EEEEEE;}\r\nfieldset{padding:0 15px 10px 15px;} \r\nh1{font-size:2.4em;margin:0;color:#FFF;}\r\nh2{font-size:1.7em;margin:0;color:#CC0000;} \r\nh3{font-size:1.2em;margin:10px 0 0 0;color:#000000;} \r\n#header{width:96%;margin:0 0 0 0;padding:6px 2% 6px 2%;font-family:\"trebuchet MS\", Verdana, sans-serif;color:#FFF;\r\nbackground-color:#555555;}\r\n#content{margin:0 0 0 2%;position:relative;}\r\n.content-container{background:#FFF;width:96%;margin-top:8px;padding:10px;position:relative;}\r\n-->\r\n</style>\r\n</head>\r\n<body>\r\n<div id=\"header\"><h1>Server Error</h1></div>\r\n<div id=\"content\">\r\n <div class=\"content-container\"><fieldset>\r\n  <h2>404 - File or directory not found.</h2>\r\n  <h3>The resource you are looking for might have been removed, had its name changed, or is temporarily unavailable.</h3>\r\n </fieldset></div>\r\n</div>\r\n</body>\r\n</html>\r\n", headers: [{"content-type", "text/html"}, {"x-docusign-node", "DA103DFE28"}, {"date", "Wed, 24 Aug 2022 20:44:25 GMT"}, {"content-length", "1245"}, {"set-cookie", "BIGipDocuSign_Demo=!gb006ncvYifkCKGlGUtBI6+PACSoFuiIM8h5+vunOMf61Qp65ooeYLvmJDAHRhDiuyn+RLSN7QOnhg==; path=/; Httponly; Secure"}, {"strict-transport-security", "max-age=31536000; includeSubDomains"}], method: :get, opts: [adapter: [timeout: 30000]], query: [include: "recipients"], status: 404, url: "https://demo.docusign.net/v2.1/accounts/64144655-50d6-OBFUSCATED/templates"}}
```

un3481 commented:
> @un3481 In the IEX console, are you able to get a list of your templates in the sandbox? In the console, I'm able to establish a connection the API endpoint, but when I get the list of my templates in Docusign, I get a `404 - File or directory not found` error back. I need to spend more time and see why this is happening, Maybe the version of Poison? Or something else?
> 
> Is it working with you? I use the same private key file as with v2.0 API, and same user ID and account ID. Everything works fine with v2.0.
> 
> ```
> > export DOCUSIGN_CLIENT_ID="xxxxxxx"
> > iex -S mix
> iex> user_id = "xxxxxxx"
> iex> {:ok, connection} = DocuSign.Connection.get(user_id)
> ...
> iex> account_id = "xxxxxx"
> iex> {:ok, templates} = DocuSign.Api.Templates.templates_get_templates(connection, account_id, include: "recipients")
> ** (MatchError) no match of right hand side value: {:error, %Tesla.Env{__client__: %Tesla.Client{adapter: {Tesla.Adapter.Mint, :call, [[]]}, fun: nil, post: [], pre: [{Tesla.Middleware.BaseUrl, :call, ["https://demo.docusign.net"]}, {Tesla.Middleware.Headers, :call, [[{"authorization", "Bearer OBFUSCATED"}]]}, {Tesla.Middleware.EncodeJson, :call, [[engine: Poison]]}]}, __module__: DocuSign.Connection.Request, body: "<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\">\r\n<html xmlns=\"http://www.w3.org/1999/xhtml\">\r\n<head>\r\n<meta http-equiv=\"Content-Type\" content=\"text/html; charset=iso-8859-1\"/>\r\n<title>404 - File or directory not found.</title>\r\n<style type=\"text/css\">\r\n<!--\r\nbody{margin:0;font-size:.7em;font-family:Verdana, Arial, Helvetica, sans-serif;background:#EEEEEE;}\r\nfieldset{padding:0 15px 10px 15px;} \r\nh1{font-size:2.4em;margin:0;color:#FFF;}\r\nh2{font-size:1.7em;margin:0;color:#CC0000;} \r\nh3{font-size:1.2em;margin:10px 0 0 0;color:#000000;} \r\n#header{width:96%;margin:0 0 0 0;padding:6px 2% 6px 2%;font-family:\"trebuchet MS\", Verdana, sans-serif;color:#FFF;\r\nbackground-color:#555555;}\r\n#content{margin:0 0 0 2%;position:relative;}\r\n.content-container{background:#FFF;width:96%;margin-top:8px;padding:10px;position:relative;}\r\n-->\r\n</style>\r\n</head>\r\n<body>\r\n<div id=\"header\"><h1>Server Error</h1></div>\r\n<div id=\"content\">\r\n <div class=\"content-container\"><fieldset>\r\n  <h2>404 - File or directory not found.</h2>\r\n  <h3>The resource you are looking for might have been removed, had its name changed, or is temporarily unavailable.</h3>\r\n </fieldset></div>\r\n</div>\r\n</body>\r\n</html>\r\n", headers: [{"content-type", "text/html"}, {"x-docusign-node", "DA103DFE28"}, {"date", "Wed, 24 Aug 2022 20:44:25 GMT"}, {"content-length", "1245"}, {"set-cookie", "BIGipDocuSign_Demo=!gb006ncvYifkCKGlGUtBI6+PACSoFuiIM8h5+vunOMf61Qp65ooeYLvmJDAHRhDiuyn+RLSN7QOnhg==; path=/; Httponly; Secure"}, {"strict-transport-security", "max-age=31536000; includeSubDomains"}], method: :get, opts: [adapter: [timeout: 30000]], query: [include: "recipients"], status: 404, url: "https://demo.docusign.net/v2.1/accounts/64144655-50d6-OBFUSCATED/templates"}}
> ```

Turns out that one fix I did for function "DocuSign.Connection.new/0" was also necessary for function "DocuSign.Connection.get/1". So now I added this fix in it and it's working fine.

frahugo commented:
Yup, that fixed the previous error. But now, I get a Deserializer error and fetching the templates from Docusign:

```
> {:ok, templates} = DocuSign.Api.Templates.templates_get_templates(connection, account_id)
** (FunctionClauseError) no function clause matching in Keyword.merge/2

    The following arguments were given to Keyword.merge/2:

        # 1
        %{
          as: %DocuSign.Model.EnvelopeTemplateResults{
            endPosition: nil,
            envelopeTemplates: nil,
            folders: nil,
.
.
.
    Attempted function clauses (showing 3 out of 3):

        def merge(keywords1, []) when is_list(keywords1)
        def merge([], keywords2) when is_list(keywords2)
        def merge(keywords1, keywords2) when is_list(keywords1) and is_list(keywords2)

    (elixir 1.13.3) lib/keyword.ex:906: Keyword.merge/2
    (docusign 0.5.2) lib/docusign/deserializer.ex:16: anonymous fn/3 in DocuSign.Deserializer.deserialize/5
    (elixir 1.13.3) lib/map.ex:830: Map.update!/3
    (docusign 0.5.2) lib/docusign/model/envelope_template_results.ex:38: Poison.Decoder.DocuSign.Model.EnvelopeTemplateResults.decode/2
    (poison 5.0.0) lib/poison.ex:86: Poison.decode!/2
    (poison 5.0.0) lib/poison.ex:61: Poison.decode/2
```

un3481 commented:
> Yup, that fixed the previous error. But now, I get a Deserializer error and fetching the templates from Docusign:
> 
> ```
> > {:ok, templates} = DocuSign.Api.Templates.templates_get_templates(connection, account_id)
> ** (FunctionClauseError) no function clause matching in Keyword.merge/2
> 
>     The following arguments were given to Keyword.merge/2:
> 
>         # 1
>         %{
>           as: %DocuSign.Model.EnvelopeTemplateResults{
>             endPosition: nil,
>             envelopeTemplates: nil,
>             folders: nil,
> .
> .
> .
>     Attempted function clauses (showing 3 out of 3):
> 
>         def merge(keywords1, []) when is_list(keywords1)
>         def merge([], keywords2) when is_list(keywords2)
>         def merge(keywords1, keywords2) when is_list(keywords1) and is_list(keywords2)
> 
>     (elixir 1.13.3) lib/keyword.ex:906: Keyword.merge/2
>     (docusign 0.5.2) lib/docusign/deserializer.ex:16: anonymous fn/3 in DocuSign.Deserializer.deserialize/5
>     (elixir 1.13.3) lib/map.ex:830: Map.update!/3
>     (docusign 0.5.2) lib/docusign/model/envelope_template_results.ex:38: Poison.Decoder.DocuSign.Model.EnvelopeTemplateResults.decode/2
>     (poison 5.0.0) lib/poison.ex:86: Poison.decode!/2
>     (poison 5.0.0) lib/poison.ex:61: Poison.decode/2
> ```

The deserializer.ex file was using syntax from Poison version 3.0, I just fixed it now and I was able to get the templates correctly.

neilberkman commented:
@frahugo would you have time to test again after the latest fixes?

frahugo commented:
@neilberkman Yes, will try next week (busy these days). Cheers.

neilberkman commented:
> @neilberkman Yes, will try next week (busy these days). Cheers.

No hurry, really appreciate it!

frahugo commented:
Bad news. There seems to be a map to model conversion issue (Poison maybe?). Inner maps in a base struct are not converted to structs. Here's how to reproduce in the console.

```
> user_id = "...."
> account_id = "..."
> {:ok, connection} = DocuSign.Connection.get(user_id) 
> DocuSign.Api.UserProfiles.user_profile_get_profile(connection, account_id, user_id)
```

In version 0.4.4, we get something like this: 

```elixir
%DocuSign.Model.UserProfiles{
   address: %DocuSign.Model.AddressInformationV2{
     address1: "",
     address2: "",
     city: "",
     country: "",
     fax: "",
     phone: "", 
     postalCode: "",
     stateOrProvince: ""
   },
   authenticationMethods: nil,
   companyName: nil,
   displayOrganizationInfo: "false",
   displayPersonalInfo: "false",
   displayProfile: "true",
   displayUsageHistory: "false",
   profileImageUri: nil,
   title: nil,
   usageHistory: %DocuSign.Model.UsageHistory{
     lastSentDateTime: "2022-05-16T17:54:36.1500000Z",
     lastSignedDateTime: "2022-05-16T17:52:23.5970000Z",
     sentCount: 65,
     signedCount: 11
   },
   userDetails: %DocuSign.Model.Users{
     lastName: "Frappier",
     workAddress: nil,
     groupList: nil,
     userProfileLastModifiedDate: nil,
     firstName: "Hugo",
     customSettings: nil,
     userId: "xxxxxxxxxxxxxxx-xxxxxxxxxx-xxxxxxxxxxxxxx",
     userType: nil,
     passwordExpiration: nil,
     userSettings: nil,
     middleName: nil,
     errorDetails: nil,
     enableConnectForUser: nil,
     forgottenPasswordInfo: nil,
     initialsImageUri: nil,
     sendActivationEmail: nil,
     uri: "/users/xxxxxxxxxxxxxxx-xxxxxxxxxx-xxxxxxxxxxxxxx",
     countryCode: nil,
     activationAccessCode: nil,
     profileImageUri: nil,
     password: nil,
     sendActivationOnInvalidLogin: nil,
     permissionProfileName: "Account Administrator",
     subscribe: nil,
     userStatus: nil,
     permissionProfileId: "xxxxxxxxxxxxxxx-xxxxxxxxxx-xxxxxxxxxxxxxx",
     userName: "Hugo Frappier",
     lastLogin: "2022-04-12T14:02:33.9470000Z",
     homeAddress: nil,
     loginStatus: nil,
     createdDateTime: "2021-04-19T20:35:35.0170000Z",
     signatureImageUri: nil,
     isAdmin: nil,
     email: "xxxxxxxxxxxxxxx-xxxxxxxxxx-xxxxxxxxxxxxxx",
     accountManagementGranular: nil,
     title: nil,
     suffixName: nil
   },
   userProfileLastModifiedDate: "2021-04-19T13:36:20.8700000Z"
 }}
```

In this PR, we get this:

```elixir
 %DocuSign.Model.UserProfile{
   address: %{
     "address1" => "",
     "address2" => "",
     "city" => "",
     "country" => "",
     "fax" => "",
     "phone" => "",
     "postalCode" => "",
     "stateOrProvince" => ""
   },
   authenticationMethods: nil,
   companyName: nil,
   displayOrganizationInfo: "false",
   displayPersonalInfo: "false",
   displayProfile: "true",
   displayUsageHistory: "false",
   profileImageUri: nil,
   title: nil,
   usageHistory: %{
     "lastSentDateTime" => "2022-05-16T17:54:36.1500000Z",
     "lastSignedDateTime" => "2022-05-16T17:52:23.5970000Z",
     "sentCount" => "65",
     "signedCount" => "11"
   },
   userDetails: %{
     "createdDateTime" => "2021-04-19T20:35:35.0170000Z",
     "defaultAccountId" => "xxxxxxxxxxxxxxx-xxxxxxxxxx-xxxxxxxxxxxxxx",
     "email" => "xxxxxxxxxxxxxxx-xxxxxxxxxx-xxxxxxxxxxxxxx",
     "firstName" => "Hugo",
     "lastLogin" => "2022-04-12T14:02:33.9470000Z",
     "lastName" => "Frappier",
     "permissionProfileId" => "xxxxxxxxxxxxxxx-xxxxxxxxxx-xxxxxxxxxxxxxx",
     "permissionProfileName" => "Account Administrator",
     "uri" => "/users/xxxxxxxxxxxxxxx-xxxxxxxxxx-xxxxxxxxxxxxxx",
     "userAddedToAccountDateTime" => "0001-01-01T08:00:00.0000000Z",
     "userId" => "56413996-fddc-4f7c-b67e-9b6f92644964",
     "userName" => "Hugo Frappier"
   },
   userProfileLastModifiedDate: "2021-04-19T13:36:20.8700000Z"
 }}
```

Note: this is happening with all models, not just the `UserProfile` one.

neilberkman commented:
> @neilberkman Bad news. There seems to be a map to model conversion issue (Poison maybe?). Inner maps in a base struct are not converted to structs. Here's how to reproduce in the console.

Thanks @frahugo 👍

un3481 commented:
Hello,

@neilberkman, I generated the code again using openapi-generator v6.2.1 and, after a couple more adjustments, I managed to fix the latest issue regarding map to model conversion.

The issue happened because openapi was using Poison v3 in the generated code while your app originally used v5. So I had to properly implement "deserialization.ex" with Poison v5.

@frahugo, when you have, please test this latest commit and let me know what you think.

Best Regards

neilberkman commented:
> @neilberkman Bad news.[...]

@frahugo I'm hopeful that @un3481's last update will have solved your problem. Whenever you have some time, it would be great if you could check this again. Thanks!

frahugo commented:
@neilberkman Yes, I had time to test it out. It is getting better. But now, I do get an error when posting an envelope.

In version 0.4.4, we get an OK response:

```elixir
> user_id = "56413996-fddc-4f7c-b67e-^^^^^^^^^^"
> account_id = "64144655-50d6-4987-853a-^^^^^^^^^^""
> template_id = "47bc3008-0ed4-4718-99ef-^^^^^^^^^^""
> {:ok, connection} = DocuSign.Connection.get(user_id)
> definition = %DocuSign.Model.EnvelopeDefinition{status: "created", templateId: template_id}
> {:ok, env} = DocuSign.Api.Envelopes.envelopes_post_envelopes(connection, account_id, envelopeDefinition: definition)
{:ok,
 %DocuSign.Model.EnvelopeSummary{
   envelopeId: "2c4270ee-5d9d-437c-8c82-^^^^^^^^^^"",
   status: "created",
   statusDateTime: "2022-11-15T20:58:28.9200000Z",
   uri: "/envelopes/2c4270ee-5d9d-437c-8c82-^^^^^^^^^^""
 }}
```

With this PR, we get an error: 

```elixir
> user_id = "56413996-fddc-4f7c-b67e-^^^^^^^^^^"
> account_id = "64144655-50d6-4987-853a-^^^^^^^^^^""
> template_id = "47bc3008-0ed4-4718-99ef-^^^^^^^^^^""
> {:ok, connection} = DocuSign.Connection.get(user_id)
> definition = %DocuSign.Model.EnvelopeDefinition{status: "created", templateId: template_id}
> {:ok, env} = DocuSign.Api.Envelopes.envelopes_post_envelopes(connection, account_id, envelopeDefinition: definition)
{:ok,
 %DocuSign.Model.ErrorDetails{
   errorCode: "INVALID_REQUEST_BODY",
   message: "The request body is missing or improperly formatted."
 }}
```

un3481 commented:
Hello @frahugo, 

What happens is, after generating the code for this new version of the API some things have changed, including the keyword options for some functions.

So, currently, in `DocuSign.Api.Envelopes.envelopes_post_envelopes/3` you have to pass the envelope definition as `body: definition` instead of `envelopeDefinition: definition`, like the following:

```elixir
> {:ok, connection} = DocuSign.Connection.get(user_id)
> definition = %DocuSign.Model.EnvelopeDefinition{status: "created", templateId: template_id}
> {:ok, env} = DocuSign.Api.Envelopes.envelopes_post_envelopes(connection, account_id, body: definition)
```

I also want to ask you if this keyword option was ever manually changed by you after the original code was generated. If that's the case I can re-implement any changes that you may have done to that generated code in order to keep things compatible with the current API.

But, if you never made any changes to this part of the code (and that was automatically generated), I suggest we use the keyword options present in this new version of the code. That's because I believe there might be a lot of small changes like this throughout the code that may be hard to track down and change manually.

Please let me know what you think.

Best Regards.

frahugo commented:
Hello @neilberkman 

I tried with the `body:` param and it does work. The envelope got created. Thanks for the tip.

But now, when I fetch the tabs of the newly created envelope, they all have `nil` values in their struct. I get the tabs this way:

```
{:ok, tabs} = DocuSign.Api.EnvelopeRecipientTabs.recipients_get_recipient_tabs(connection, account_id, envelope_id, 1)
```

Is there something else I need to do?

Cheers

un3481 commented:
Hello @frahugo 

I finally managed to properly upgrade Poison to v5. The problem happened due to some changes in `Poison.decode/2` from v3 to v5, so I had to ignore the v5 function and implement a code similar to what was in v3.

I believe now the issues with the deserializer are all fixed. Please test the latest commit I made and let me know what you think.

Best Regards.

neilberkman commented:
> Is there something else I need to do?

Hi @frahugo as @un3481 mentioned he's updated the branch and hopefully the issues you encountered previously are resolved. If you're able to give this another try I'd really appreciate it. Thanks!

frahugo commented:
Hi @neilberkman and @un3481.  So far so good. I did a quick test in the console and was able to create an enveloppe. Next week, I'll with our application and report. Thanks a lot!

frahugo commented:
Hi @neilberkman and @un3481. I included the latest code in our app and updated the app to use the new `body:` params syntax. Works great with one of the two Envelope templates we have (a very simple one-pager). But with the second one, it breaks. It uses `Number` fields, and the number tabs are not properly deserialized.  In 7 models, this line should be changed from

```
          :numberTabs => [DocuSign.Model.Float.t()] | nil,
```

to

```
          :numberTabs => [DocuSign.Model.Number.t()] | nil,
```

I suspect the API definition provided by DocuSign is wrong, hence the error in the code generation.

Let me know when everything is ready for another round of tests.

un3481 commented:
Hello @frahugo, @neilberkman,

I will work on this issue as soon as possible. The code was generated with the latest stable version of OpenAPI Generator, but I will also take a look at their open issues to see if I find anything helpful. 

Best Regards.

frahugo commented:
Awesome. I remember I had to manually fix it in 2021 (see commit: https://github.com/neilberkman/docusign_elixir/commit/c2e2c0e154ce3c4d33adf776b36bd06ab823e8d8). Either the API definition provided by DocuSign or code generator is buggy.

> Hello @frahugo, @neilberkman,
> 
> I will work on this issue as soon as possible. The code was generated with the latest stable version of OpenAPI Generator, but I will also take a look at their open issues to see if I find anything helpful.
> 
> Best Regards.

un3481 commented:
Hello @frahugo, @neilberkman 

I tried changing some things in the Swagger file from DocuSign and I believe I understood the issue.

They have a definition called "number", which should be translated into a struct in elixir. The problem is: there is a type already in OpenAPI called "number" that is translated to the "Float" struct that you found in the bad models @frahugo.

So I just renamed all occurences of this "number" definition in the swagger file to "docuSignNumber" and the name conflict seems to be solved. You can verify that all the models you mentioned now have "DocuSignNumber" instead of "Float".

I'm thinking about creating a PR in DocuSign official repo for the Swagger file with my fixes, but first I would like you to test this latest commits and confirm to me that there are no more issues with the generated code.

Best Regards.

frahugo commented:
Excellent @un3481. I'm happy to report that this time, eveything on my side works perfectly. I think it is ready for roll-out. It might be worth mentionning that this is a breaking upgrade. I had to do a couple of changes in the app:

1. `DocuSign.Model.EnvelopeTemplateResult` renamed to `DocuSign.Model.EnvelopeTemplate`
2. `DocuSign.Model.Number` renamed to `DocuSign.Model.DocuSignNumber`
3. Use `body` as optional argument to some functions (i.e. `body` instead of `EnvelopeRecipientTabs` when calling `DocuSign.Api.EnvelopeRecipientTabs.recipients_put_recipient_tabs()`)

Cheers!

un3481 commented:
Hello @frahugo, that's great! You may want to update the README file with those breaking changes when you do your next release, maybe I could help you with that. If you need anything else just let me know.

Best Regards.

neilberkman commented:
Thanks @frahugo for testing and for the guidance on doc updates! And thanks @un3481 for the PR! I'll update the README and do a new release soon.

frahugo commented:
@un3481 @neilberkman Bad news. Our CI reported some missing modules, thanks to dialyzer. Not sure if the list is complete, but these 2 modules are missing but used in some type specs:

```
DocuSign.Model.DateTime
DocuSign.Model.OfflineAttributes
```

un3481 commented:
Hello @frahugo @neilberkman,

I checked DocuSign's official repo and I found an open [issue](https://github.com/docusign/OpenAPI-Specifications/issues/36) regarding `OfflineAttributes` type. That appears to be a mistake from them. So I just removed this type definition from the Swagger file. I believe I can't do much more than that until they close this open issue.

Regarding the `Date` type, there was a bug exactly like the one with `Number`. There is a type already in OpenAPI specs called "date". So I renamed it to `DocuSignDate` and now I believe this conflict is resolved.

Please test this latest commits again when you have the time and let me know if everything works and if the CI is passing.

Best Regards.

frahugo commented:
Hello @un3481. Everything is green on my side. So we can proceed with the merge.

neilberkman commented:
> Hello @un3481. Everything is green on my side. So we can proceed with the merge.

Thanks @frahugo, great news! I'll update the `README` and do the merge soon.

neilberkman commented:
@un3481 and @frahugo thanks for all of your help with this! I just published v1.0.0 to hex.pm.
